Fix missing case for function pointer signatures in Crossgen2

trylek opened the pull request:
As Andy Ayers conjectured in the issue #72822 and I confirmed,
Crossgen2 signature emitter was missing the code path for
emitting function pointer signatures. This change fixes this
deficiency. The change also removes the issues.targets exclusion
of the affected test that Andy originally merged in to mitigate
the issue.

Thanks

Tomas

P.S. Even though MethodSignatureFlags more or less corresponds
to CoreCLR runtime calling convention, I have ignored this fact
and implemented an explicit conversion as I find it cleaner, please
let me know if you think otherwise. I would also appreciate
double-checking of the logic around Static vs. HASTHIS as I'm not
completely sure about the exact interpretation of this runtime flag.

/cc @dotnet/crossgen-contrib 

Fixes: https://github.com/dotnet/runtime/issues/72822

trylek committed:
Fix missing case for function pointer signatures in Crossgen2

As Andy Ayers suggested in the issue #72822 and I confirmed,
Crossgen2 signature emitter was missing the code path for
emitting function pointer signatures. This change fixes this
deficiency. The change also removes the issues.targets exclusion
of the affected test that Andy originally merged in to mitigate
the issue.

Thanks

Tomas

jkotas commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs:
```diff
@@ -366,6 +370,35 @@ private void EmitPointerTypeSignature(PointerType type, SignatureContext context
+            {
+                callingConvention |= CorCallingConvention.IMAGE_CEE_CS_CALLCONV_HASTHIS | CorCallingConvention.IMAGE_CEE_CS_CALLCONV_EXPLICITTHIS;
+            }
+
```
Does this need to emit modops / modreqs, or are those intentionally skipped here?

trylek replied:
Well, we probably do, I'll try to fix this but the problem is that apparently there's no pre-existing coverage for function pointer signatures in CoreCLR testing so I guess I'll need to create some dedicated tests to verify this works as expected especially as this is the first place in Crossgen2 where we'll be emitting these signature modifiers.

jkotas replied:
Actually - if we are ever only going to use signature to create type handle at runtime, we should not need to emit the modopts/modreqs. They are not part of the type handle at runtime.

jkotas replied:
Also, the function pointer support in reflection is broken and inconsistent in various ways today. We hope to fix that as part of https://github.com/dotnet/runtime/issues/69273 in .NET 8.

trylek replied:
That's exactly why I asked for a broad review, this is obviously new functionality with very limited test coverage (amounting to the one JIT test I'm re-enabling in this PR).

davidwrighton replied:
We have logic to handle emission of modopts/modreqs, but that should only be needed for function signatures, and those are already handled correctly so far as I'm aware. For this support I do not believe we need to handle modreqs/modopts at all, ever.

jkotas commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs:
```diff
@@ -366,6 +370,35 @@ private void EmitPointerTypeSignature(PointerType type, SignatureContext context
+
+            if ((type.Signature.Flags & MethodSignatureFlags.Static) == 0)
+            {
+                callingConvention |= CorCallingConvention.IMAGE_CEE_CS_CALLCONV_HASTHIS | CorCallingConvention.IMAGE_CEE_CS_CALLCONV_EXPLICITTHIS;
```
Is `IMAGE_CEE_CS_CALLCONV_EXPLICITTHIS` right here?

trylek replied:
Hmm, that's exactly what I'm trying to get my head around. I believe that ECMA managed method signatures do include the <code>this</code> pointer so I would assume it's explicit but I may be misunderstanding the purpose of the flag, it's not used very broadly in the CoreCLR native runtime.

jkotas replied:
The MethodSignature initialization only uses `HasThis` to initialize the `Static` bit:

https://github.com/dotnet/runtime/blob/83bd401b112f2deb0db516300f72eb9d1579751d/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaSignatureParser.cs#L399-L400

so the reverse should do the same.

If we need to preserve HasExplicitThis too (I do not think we need to), we would need to store it in MethodSignature.

trylek replied:
Based on your clarification I have removed the <code>ExplicitThis</code> flag when emitting the non-static function pointer signatures.

jkotas replied:
I am still not sure about the ExplicitThis. I think we should probably store it on MethodSignature when we are initializing it from Ecma metadata, store it here; and also send it to the JIT so that the JIT can do the right thing with it.

jkotas replied:
We have number of IL tests that exercise explicit this (look for "explicit instance" in *.il under tests). I am surprised that none of them fails.

davidwrighton replied:
Ack, yes, we should handle explicitthis, although fixing all the places that work with that may be beyond the scope of what we should try to fix with this PR. I think it will likely be a fairly large fix to a bunch of places unrelated to the general thrust of the code written here.

trylek replied:
I have created a .NET 8 tracking issue for this task, https://github.com/dotnet/runtime/issues/73407.

jkotas commented on src/coreclr/tools/Common/Internal/Runtime/CorConstants.cs:
```diff
@@ -95,4 +95,36 @@ public enum CorTokenType
     }
+
+    [Flags]
+    public enum CorUnmanagedCallingConvention
```
These constants are defined by System.Reflection.Metadata as `SignatureCallingConvention` and `SignatureAttributes`. We can just use those definitions instead of having yet another copy.

trylek replied:
Thanks Jan for pointing that out, I have reverted the additions to CorConstants and modified the code to reuse the pre-existing constants.

jkotas commented:
>  I have ignored this fact
and implemented an explicit conversion as I find it cleaner,

FWIW, the type system takes the casting shortcut:

https://github.com/dotnet/runtime/blob/83bd401b112f2deb0db516300f72eb9d1579751d/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaSignatureParser.cs#L389-L394

trylek committed:
Address JanK's PR feedback

I have reverted the additions to CorConstants and modified the
code to use the constants in System.Reflection.Metadata. I have
removed the flag translation based on Jan's confirmation that
the calling convention part of MethodSignatureFlags is indeed
kept in sync with the CoreCLR native runtime calling conventions
and I dropped the explicit this flag.

Thanks

Tomas

davidwrighton reviewed (approved):
I believe we should probably fix the explicit this problem, but I'd like to see such a fix done as part of fixing explicit this throughout our product, not just in this situation.

trylek commented:
/azp run runtime-coreclr crossgen2
